src/dattobd.c

```diff
@@ -4229,6 +4311,52 @@ static long ctrl_ioctl(struct file *filp, unsigned int cmd, unsigned long arg){
 		}
 
 		break;
+    case IOCTL_DATTOBD_VERSION:
+        src = DATTOBD_VERSION;
+        dest = ((struct dattobd_version __user *)arg)->version_string;
+        len = strlen(src) + 1;
+        vslen = sizeof((struct dattobd_version *)0)->version_string;
+        // make sure we don't copy more than the user should have allocated.
+        if (len > vslen)
+            len = vslen;
+        ret = copy_to_user(dest, src, len);
+        // if the string is too long, make sure to nt the last position of the user's buffer.
+        if ((ret == 0) && (len == vslen))
+            ret = copy_to_user(dest+len-1, "", 1);
+        if(ret){
+            ret = -EFAULT;
+            LOG_ERROR(ret, "error copying dattobd version to user space");
+            break;
+        }
+
+        break;
+    case IOCTL_DATTOBD_ACTIVE_DEVICE_INFO:
+        //get the struct from user space to get the count of memory they're providing
+        adinfo = kmalloc(sizeof(struct dattobd_active_device_info), GFP_KERNEL);
+        if(!adinfo){
+            ret = -ENOMEM;
+            LOG_ERROR(ret, "error allocating memory for dattobd all device info");
+            break;
+        }
+
+        ret = copy_from_user(adinfo, (struct dattobd_active_device_info __user *)arg, sizeof(struct dattobd_active_device_info));
+        if(ret){
+            ret = -EFAULT;
+            LOG_ERROR(ret, "error copying dattobd all device info struct from user space");
+            break;
+        }
+
+        copyback = 0;
+        ret = ioctl_dattobd_active_device_info(&adinfo, &copyback);
+        if(ret) break;
+
+        ret = copy_to_user((struct dattobd_info __user *)arg, adinfo, copyback);
```

Comment:
I would prefer it if copyback was instead the number of items in the dattobd_info array. This way, the `ctrl_ioctl()` function handles figuring out how much data to copy to userspace and the handler function handles just what it needs to to populate the adinfo struct.